base/shell/explorer/traywnd.cpp

```diff
@@ -607,6 +662,9 @@ class CTrayWindow :
             break;
 
         case ID_SHELL_CMD_UNDO_ACTION:
+            m_bTiled = FALSE;
+            m_bCascaded = FALSE;
+            RestoreWindowPos();
```

Comment:
How long does this take if you have like 15 or 20 windows open?
You should probably not block the thread that long....

Reply:
Quite fast but i'll optimize by using BeginDeferWindowPos / DeferWindowPos / EndDeferWindowPos

Reply:
Best option would be even to call NtUserEndDeferWindowPosEx to do it async but I don't know how to call this Win32ss function from here :-(

Reply:
In general, it should not be required to call functions like that from the shell.
I was more thinking about doing this restoring on a thread.

Reply:
OK. Defer processing does not drastically change the situation (which is quite good anyway)
Regarding your feared scenario, I understand the principle but :
- When you open 20+ windows, quite unlikely to tile/cascade (too cluttered) and therefore even less likely to "undo" the operation
- Having a dedicated thread for "restore" only make sense if Tile / Cascade is refactored in the same way.

I propose to have this first implementation in a single thread approach (simpler, robust, fit for purpose for 99% of ReactOS use cases) and to log in code comments a suggestion for future improvement. OK for you ?